moveit_core/collision_detection_fcl/src/collision_common.cpp

```diff
@@ -463,37 +462,141 @@ bool distanceCallback(fcl::CollisionObject* o1, fcl::CollisionObject* o2, void*
+    {
+      // If at the limit for a given pair just return
+      if (it->second.size() >= cdata->req->max_contacts_per_body)
+      {
+        return cdata->done;
+      }
+    }
+    else if (cdata->req->type != DistanceRequestType::GLOBAL)
+    {
+      dist_threshold = cdata->res->minimum_distance.distance;
+    }
+    else if (cdata->req->type != DistanceRequestType::SINGLE)
+    {
+      dist_threshold = it->second[0].distance;
+    }
   }
-  else
+
+  fcl_result.min_distance = dist_threshold;
+  double d = fcl::distance(o1, o2, fcl::DistanceRequest(cdata->req->enable_nearest_points), fcl_result);
+
+  // Check if either object is already in the map. If not add it or if present
+  // check to see if the new distance is closer. If closer remove the existing
+  // one and add the new distance information.
+  if (d < dist_threshold)
   {
-    if (cdata->res_->distance > d)
+    dist_result.distance = fcl_result.min_distance;
+    dist_result.nearest_points[0] = Eigen::Vector3d(fcl_result.nearest_points[0].data.vs);
+    dist_result.nearest_points[1] = Eigen::Vector3d(fcl_result.nearest_points[1].data.vs);
+    dist_result.link_names[0] = cd1->ptr.obj->id_;
+    dist_result.link_names[1] = cd2->ptr.obj->id_;
+    dist_result.body_types[0] = cd1->type;
+    dist_result.body_types[1] = cd2->type;
+    if (cdata->req->enable_nearest_points)
     {
-      if (cdata->req_->verbose)
-        CONSOLE_BRIDGE_logWarn("Distance between %s and %s: %f decreased", cd1->getID().c_str(), cd2->getID().c_str(),
-                               d);
-      cdata->res_->distance = d;
+      dist_result.normal = (dist_result.nearest_points[1] - dist_result.nearest_points[0]).normalized();
     }
-  }
 
-  min_dist = cdata->res_->distance;
+    if (d <= 0 && cdata->req->enable_signed_distance)
+    {
+      dist_result.nearest_points[0].setZero();
+      dist_result.nearest_points[1].setZero();
+      dist_result.normal.setZero();
+
+      fcl::CollisionRequest coll_req;
+      fcl::CollisionResult coll_res;
+      coll_req.enable_contact = true;
+      coll_req.num_max_contacts = 200;
+      std::size_t contacts = fcl::collide(o1, o2, coll_req, coll_res);
+      if (contacts > 0)
+      {
+        double max_dist = 0;
+        int max_index = 0;
+        for (int i = 0; i < contacts; ++i)
+        {
+          const fcl::Contact& contact = coll_res.getContact(i);
+          if (contact.penetration_depth > max_dist)
+          {
+            max_dist = contact.penetration_depth;
+            max_index = i;
+          }
+        }
 
-  return cdata->done_;
+        const fcl::Contact& contact = coll_res.getContact(max_index);
+        dist_result.distance = -contact.penetration_depth;
+        dist_result.nearest_points[0] = Eigen::Vector3d(contact.pos.data.vs);
+        dist_result.nearest_points[1] = Eigen::Vector3d(contact.pos.data.vs);
+        dist_result.normal = Eigen::Vector3d(contact.normal.data.vs);
```

Comment:
Here you deviate from your definition that the normal vector connects nearest points. This should be documented.

Reply:
This is a deviation. If you are using convex to convex checking and in contact it would be as documented but in the case of non-convex to non-convex fcl provides the contact normal (One of the triangle normals). From what I could tell the way it choose which triangle normal to use it picked the one with the least penetration. Also for non-convex to non-convex if the objects are not in contact it is as documented. I something along what I just described sufficient to add?

In the case of moveit there is not a way to tell fcl to use convex representation so currently everything uses non-convex but it would be nice to support this in the future.

Reply:
Yes, such an explanation will be fine.